### PR TITLE
[Android] MapBuffer long support

### PR DESCRIPTION
**Pull request by ryanlntn**
## Summary:

This adds support for 64 bit integer (long) values to MapBuffer. Per the wide gamut color [RFC](https://github.com/react-native-community/discussions-and-proposals/pull/738) Android encodes wide gamut colors as long values so we need to update MapBuffer to support 64 bit integers as well.

## Changelog:

[ANDROID] [ADDED] - Add 64 bit integer (long) value support to MapBuffer

## Test Plan:

I've added a test to the MapBuffer test suite. This new API is otherwise currently unused but will be used in subsequent PRs as part of wide gamut color support changes.

### CODE REVIEW

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/bridge/JavaOnlyArray.java [hunk 1]
```diff
@@ -156,6 +161,11 @@ public void pushInt(int value) {
     mBackingList.add(new Double(value));
   }
 
+  @Override
+  public void pushLong(long value) {
+    mBackingList.add(new Double(value));
```
Comment (NickGerleman) [inline packages/react-native/ReactAndroid/src/main/java/com/facebook/react/bridge/JavaOnlyArray.java:166]: It's not ideal that the `long` API stores a double, and doesn't support values > 2^53. It is the only lossy option here.

Would it be too complicated to store as `Long` here, or to instead, use existing `double` API to make this truncation explicit?

If we are modeling JS number that's a double under the hood, and has that limitation, it might be best to model that until we can safely convert to longish.

File: packages/react-native/ReactCommon/react/renderer/mapbuffer/MapBufferBuilder.cpp [hunk 2]
```diff
@@ -15,6 +15,7 @@ namespace facebook::react {
 constexpr uint32_t INT_SIZE = sizeof(uint32_t);
 constexpr uint32_t DOUBLE_SIZE = sizeof(double);
 constexpr uint32_t MAX_BUCKET_VALUE_SIZE = sizeof(uint64_t);
+constexpr uint64_t LONG_SIZE = sizeof(uint64_t);
```
Comment (NickGerleman) [inline packages/react-native/ReactCommon/react/renderer/mapbuffer/MapBufferBuilder.cpp:18]: nits:
1. This be typed `uint32_t`
2. It should be placed next to int/double sizes

File: packages/react-native/ReactCommon/react/renderer/mapbuffer/tests/MapBufferTest.cpp [hunk 3]
```diff
@@ -27,6 +28,22 @@ TEST(MapBufferTest, testSimpleIntMap) {
   EXPECT_EQ(map.getInt(1), 4321);
 }
 
+TEST(MapBufferTest, testSimpleLongMap) {
+  auto builder = MapBufferBuilder();
+
+  int64_t minInt64 = std::numeric_limits<int64_t>::min();
+  int64_t maxInt64 = std::numeric_limits<int64_t>::max();
+
+  builder.putLong(0, minInt64);
```
Comment (NickGerleman) [inline packages/react-native/ReactCommon/react/renderer/mapbuffer/tests/MapBufferTest.cpp:37]: Could we add tests for more interesting cases? Like large (>52 bit) numbers, negative numbers?

**Review by NickGerleman — COMMENTED** [review]
Ooh this is great!

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/bridge/JavaOnlyArray.java (hunk 1 above)
Comment (NickGerleman) [reply packages/react-native/ReactAndroid/src/main/java/com/facebook/react/bridge/JavaOnlyArray.java:166]: Hmm, this would shave off the color space, and part of the alpha, if we tried to put an android color long through it.

image

Comment (ryanlntn) [reply packages/react-native/ReactAndroid/src/main/java/com/facebook/react/bridge/JavaOnlyArray.java:166]: Oh good catch. I think I accidentally brought that over from pushInt.

**Review by NickGerleman — APPROVED** [review]

**Comment by ryanlntn** [comment]
@NickGerleman I dropped the bridge changes to keep this completely focused on MapBuffer.
